### ISSUE TITLE
.type() misbehaves when input's value is changed programatically

### ISSUE THREAD

**Issue by kamituel** [cypress-io/cypress#568]
- Operating System:
Mac OS X 10.11.6

- Cypress Version:
0.19.4

- Browser Version:
Canary 61

### Is this a Feature or Bug?
I think it's a bug. It's possible I'm just not simulating a tab key in a way that lets Cypress pick caret position up, though.

### Current behavior:
When, for the `<input>`, I simulate pressing a tab key, and then type into the field, and then do it again, `.type()` will not append a text at the end of the input value, but in the middle.

Test scenario pasted below fails with the following error:

> AssertionError: expected '<input>' to have value 'autocomplete123autocomplete456autocomplete', but the value was 'autocomplete456123autocompleteautocomplete'

### Desired behavior:
I would expect `.type()` to always insert a text at the end of the input (or maybe, more generally, wherever the caret is).

### How to reproduce:
Serve this file:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  </head>
  <body>
    <input>

    <script>
      var input = document.querySelector("input");

      input.addEventListener("keydown", function(evt) {
      	if (evt.key === "Tab") {
          evt.preventDefault();
          input.value += "autocomplete";
        }
      });
    </script>
  </body>
</html>
```

Then execute test.

#### Test code:

```js
describe("Quirk", () => {
  beforeEach(() => {
    cy.visit("http://localhost:8000/cypress-type-quirk.html");
  });

  Cypress.addChildCommand("tab", ($subj) => {
    let eDown = new KeyboardEvent("keydown", {key: "Tab", bubbles: true});
    $subj[0].dispatchEvent(eDown);

    let log = Cypress.Log.command({name: "tab"});

    cy.chain()
      .wrap($subj, {log: false})
      .then(() => log.snapshot().end());
  });

  it("misbehaves", () => {
    cy
      .get("input")
        .tab().should("have.value", "autocomplete")
        .type("123").tab().should("have.value", "autocomplete123autocomplete")
        // This one fails.
        .type("456").tab().should("have.value", "autocomplete123autocomplete456autocomplete")
  });
});
```

### Additional Info (images, stack traces, etc)

Just to be clear: at the last step, input actually has an invalid value `"autocomplete456123autocompleteautocomplete"` - I can see it in the browser window.

Also, the caret (the blinking text cursor) disappears after `.type("456")` is invoked. Not sure how it relates to the issue though.

Here's a screencast of this happening: https://drive.google.com/open?id=0B_Kr8lR7GpI6VDVPeHB4SWQzZm8

**Comment by kamituel** [comment]
Using this command instead of `.type()` solves this for me, if it helps:

```js
  Cypress.addChildCommand("typee", ($subj, text) => {
    $subj[0].value += text;
    $subj[0].dispatchEvent(new Event("input", {bubbles: true}));

    cy.chain()
      .wrap($subj, {log: false});
  });
```

**Comment by chrisbreiding** [comment]
Thank you for the issue and repro. This will be fixed once we release `0.20.0`. We'll update this issue and reference the changelog when it's released.

**Comment by kamituel** [comment]
Good to know, thanks! One more reason to be excited about `0.20.0`!

**Comment by brian-mann** [comment]
Fixed in [`0.20.0`](https://docs.cypress.io/guides/references/changelog.html#0-20-0)
